### PR TITLE
feat: adding history loading state

### LINKED ISSUE #382: History View > Loading State and Transition between commits
Instead of a spinner, the body shows "There's nothing here yet" until the body loads.  A spinner would be good, but maybe we can show a grayed out table.

### PR DESCRIPTION
**Pull request by boandriy**
adding history loading state

closes #382

### CODE REVIEW

**Review by chriswhong — CHANGES_REQUESTED** [review]
- The loading content for `CommitSummaryHeader` isn't the same height as the content, so there is a shift that happens when it appears and disappears.  Let's make sure this element has a constant height
- The tab for "Data" is wider than the other tabs by design, but during the loading state, it's the same width.  Let's make sure the Data tab is always the same width and is wider than the others.
- The body still displays "There's nothing here yet" during part of the loading... When I tested it I saw the new table with animated loading rectangles in it, but this went away, then it shows "There's nothing here yet", then the body renders.  Let's make sure the loading state table shows continuously until the body is in place.

**Comment by boandriy** [comment]
Hey @chriswhong , There is still a content shift, as the header is missing on the loading state. Should I add similar content loader?

**Review by chriswhong — CHANGES_REQUESTED** [review]
Still seeing a very slight shift in content when switching between versions in history.  Also, the heading Version Info can stay as it doesn't change, only the content below it changes.

Can we resize/reformat the SVG so that it matches the height of the content it is replacing and prevents the shift completely?

**Comment by chriswhong** [comment]
A few more changes:
- The active tab shouldn't be changing either.  I am seeing the data tab as white, then choosing another version, and it flashes gray and then white again.  Let's make sure this flash doesn't occur, the active tab should stay active through the loading state.
- I see "showing all 0 rows" for a moment which is awkward.  I think because these body header icons and text are derived from the content that is loading, they should also show loading state.

File: src/features/dataset/state/datasetState.ts [hunk 1]
```diff
@@ -78,8 +84,12 @@ export const datasetReducer = createReducer(initialState, {
   'API_DATASET_FAILURE': (state) => {
     state.datasetLoading = false
   },
+  'API_BODY_REQUEST': (state, action) => {
+    state.bodyLoading = true
+  },
   'API_BODY_SUCCESS': (state, action) => {
     state.dataset.body = action.payload.data as Body
+    state.bodyLoading = false
```
Comment (ramfox) [inline src/features/dataset/state/datasetState.ts:94]: need to also track `API_BODY_FAILURE`, and set the `bodyLoading` to false

File: src/features/dsComponents/body/Body.tsx [hunk 2]
```diff
@@ -35,6 +36,8 @@ export interface BodyProps {
 // }
 
 const extractColumnHeaders = (structure: Structure, value: any[]): ColumnProperties[] => {
+
```
Comment (ramfox) [inline src/features/dsComponents/body/Body.tsx]: extra line here

File: src/features/dsComponents/body/Body.tsx [hunk 3]
```diff
@@ -1,12 +1,13 @@
...
-import { useDispatch } from 'react-redux'
+import {useDispatch, useSelector} from 'react-redux'
 // import { Action } from 'redux'
 
 import Dataset,  { Structure, schemaToColumns, ColumnProperties } from '../../../qri/dataset'
 import BodyTable from './BodyTable'
 import BodyJson from './BodyJson'
 import { loadBody } from '../../dataset/state/datasetActions'
 import { newQriRef } from '../../../qri/ref'
+import {selectIsBodyLoading} from "../../dataset/state/datasetState";
```
Comment (ramfox) [inline src/features/dsComponents/body/Body.tsx]:
```
import {selectIsBodyLoading} from "../../dataset/state/datasetState";
```
=>
```
import { selectIsBodyLoading } from "../../dataset/state/datasetState"
```

File: src/features/dsComponents/body/Body.tsx [hunk 4]
```diff
@@ -1,12 +1,13 @@
 import React, { useEffect } from 'react'
-import { useDispatch } from 'react-redux'
+import {useDispatch, useSelector} from 'react-redux'
```
Comment (ramfox) [inline src/features/dsComponents/body/Body.tsx]:
```
import {useDispatch, useSelector} from 'react-redux'
```
=>
```
import { useDispatch, useSelector } from 'react-redux'
```

File: src/features/dsComponents/body/BodyTable.tsx [hunk 5]
```diff
@@ -14,6 +15,7 @@ import DataType from '../../../chrome/DataType'
 interface BodyTableProps {
   headers?: any[]
   body: any[]
+  loading: boolean
```
Comment (ramfox) [inline src/features/dsComponents/body/BodyTable.tsx]: can we switch this to optional? Then default to `false` when you destructure it from `props`? Then we can remove `loading={false}` from the `<BodyTable>` call in `Body.tsx`

File: src/features/dsComponents/body/Body.tsx [hunk 6]
```diff
@@ -125,6 +138,7 @@ const Body: React.FC<BodyProps> = ({
   return (structure.format === 'csv' && Array.isArray(body))
     ? <BodyTable
         headers={extractColumnHeaders(structure, body)}
+        loading={false}
```
Comment (ramfox) [inline src/features/dsComponents/body/Body.tsx]: Can we make the `loading` param optional (defaulting to `false`), then remove this line?

**Review by ramfox — APPROVED** [review]
A few nitpicks, but this looks great!

File: src/features/commits/CommitSummaryHeader.tsx [hunk 7]
```diff
@@ -2,31 +2,46 @@ import React from 'react'
 
 import Dataset from '../../qri/dataset'
 import DatasetVersionInfo from '../../chrome/DatasetCommitInfo'
+import ContentLoader from "react-content-loader";
```
Comment (b5) [inline src/features/commits/CommitSummaryHeader.tsx]: move package imports above local imports, separated by a blank linke

File: src/features/commits/CommitSummaryHeader.tsx [hunk 8]
```diff
@@ -2,31 +2,46 @@ import React from 'react'
...
 
 export interface CommitSummaryHeaderProps {
   dataset: Dataset
+  loading: boolean
```
Comment (b5) [inline src/features/commits/CommitSummaryHeader.tsx]: is this prop _required_? we often do loading props as optional with a default value of `false` within the component itself

File: src/features/commits/CommitSummaryHeader.tsx [hunk 9]
```diff
@@ -2,31 +2,46 @@ import React from 'react'
...
+              <div className='text-sm text-qrigray-400 font-semibold mb-2'>Version Info</div>
+              <ContentLoader width="421" height="48.6"  fill="none">
+                <rect y="4" width="421" height="12" fill="#D5DADD"/>
+                <rect y="32" width="89" height="12"  fill="#D5DADD"/>
+                <rect x="99" y="32" width="89" height="12" fill="#D5DADD"/>
+              </ContentLoader>
+            </>:
+            <>
+              <div className='text-sm text-qrigray-400 font-semibold mb-2'>Version Info</div>
+              < DatasetVersionInfo dataset={dataset} />
```
Comment (b5) [inline src/features/commits/CommitSummaryHeader.tsx]: extraneous space: `< DatasetVersion...`

File: src/features/commits/DatasetCommit.tsx [hunk 10]
```diff
@@ -6,17 +6,20 @@ import { LogItem } from '../../qri/log'
 import { newQriRef, refParamsFromLocation } from '../../qri/ref'
 import { pathToDatasetHistory } from '../dataset/state/datasetPaths'
 import DatasetCommitInfo from '../../chrome/DatasetCommitInfo'
+import ContentLoader from "react-content-loader";
```
Comment (b5) [inline src/features/commits/DatasetCommit.tsx]: move package imports above local imports

File: src/features/commits/DatasetCommitListItem.tsx [hunk 11]
```diff
@@ -6,13 +6,15 @@ import DatasetCommit from './DatasetCommit'
 
 export interface DatasetCommitListItemProps {
   logItem: LogItem
+  loading: boolean
```
Comment (b5) [inline src/features/commits/DatasetCommitListItem.tsx]: make this optional, with a default value of `false`

File: src/features/dataset/DatasetHeader.tsx [hunk 12]
```diff
@@ -8,9 +8,10 @@ import DatasetInfoItem from './DatasetInfoItem'
 import DownloadDatasetButton from '../download/DownloadDatasetButton'
 import Link from '../../chrome/Link'
 import { validateDatasetName } from '../session/state/formValidation'
-import { selectDatasetHeader } from "./state/datasetState";
+import { selectDatasetHeader, selectIsHeaderLoading } from "./state/datasetState";
 import { qriRefFromVersionInfo } from "../../qri/versionInfo";
 import fileSize from "../../utils/fileSize";
+import ContentLoader from "react-content-loader";
```
Comment (b5) [inline src/features/dataset/DatasetHeader.tsx]: move package import above local import

File: src/features/dsComponents/body/BodyTable.tsx [hunk 13]
```diff
@@ -1,5 +1,6 @@
 import React from 'react'
 import classNames from 'classnames'
+import ContentLoader from "react-content-loader";
 // import * as _ from 'underscore'
 // import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
 // import { faSync } from '@fortawesome/free-solid-svg-icons'
```
Comment (b5) [inline src/features/dsComponents/body/BodyTable.tsx]: remove these commented-out imports

File: src/features/dsComponents/body/Body.tsx [hunk 14]
```diff
@@ -60,10 +62,20 @@ const Body: React.FC<BodyProps> = ({
...
+    }
   }, [preview, dispatch, path, name, username])
 
-  if (!body) {
+  if (loading) {
+    return <BodyTable
+      headers={[]}
+      loading={true}
+      body={[]}
+    />
```
Comment (b5) [inline src/features/dsComponents/body/Body.tsx]: make `headers` and `body` BodyTable props optional, use an empty array as default values. use `loading` prop literal instead:

```ts
if (loading) {
  return <BodyTable loading />
}
```

File: src/features/dsComponents/DatasetComponents.tsx [hunk 15]
```diff
@@ -21,7 +21,10 @@ const DatasetComponents: React.FC<{}> = () => {
 
   useEffect(() => {
     const ref = newQriRef({username: qriRef.username, name: qriRef.name, path: qriRef.path})
-    dispatch(loadDataset(ref))
+    dispatch(setBodyLoading())
+    if (ref.username && ref.name) {
+      dispatch(loadDataset(ref))
+    }
```
Comment (b5) [inline src/features/dsComponents/DatasetComponents.tsx:27]: This solution works for now, but isn't optimal. Body loading state should be handled entirely within the `loadDataset` action, and a selector like `selectBodyIsLoading` should be able to work out it's state entirely from the request lifecycle of `loadDataset`.

That said, this solution will work for now, so let's merge it. Can you file an issue titled `Revisit Body Loading State on Dataset Components page` and quote my above comment?

**Review by b5 — APPROVED** [review]
LGTM

**Pull request merged**